src/Functions/LowerUpperUTF8Impl.h

```diff
@@ -37,40 +41,70 @@ struct LowerUpperUTF8Impl
             return;
         }
 
-        res_data.resize(data.size());
-        res_offsets.resize_exact(offsets.size());
 
-        String output;
+        UErrorCode error_code = U_ZERO_ERROR;
+        UCaseMap * csm = ucasemap_open(nullptr, 0, &error_code);
+        if (U_FAILURE(error_code))
+            throw DB::Exception(ErrorCodes::LOGICAL_ERROR, "Call ucasemap_open error:{}", u_errorName(error_code));
+
+        // String output;
         size_t curr_offset = 0;
+        res_data.resize(data.size());
+        res_offsets.resize_exact(offsets.size());
         for (size_t i = 0; i < input_rows_count; ++i)
         {
             const auto * data_start = reinterpret_cast<const char *>(&data[offsets[i - 1]]);
-            size_t size = offsets[i] - offsets[i - 1];
+            size_t size = offsets[i] - offsets[i - 1] - 1;
 
-            icu::UnicodeString input(data_start, static_cast<int32_t>(size), "UTF-8");
+            int32_t out_size;
             if constexpr (upper)
-                input.toUpper();
+                out_size = ucasemap_utf8ToUpper(
+                    csm, reinterpret_cast<char *>(&res_data[curr_offset]), res_data.size() - curr_offset, data_start, size, &error_code);
             else
-                input.toLower();
-
-            output.clear();
-            input.toUTF8String(output);
-
-            /// For valid UTF-8 input strings, ICU sometimes produces output with an extra '\0 at the end. Only the data before that
-            /// '\0' is valid. If the input is not valid UTF-8, then the behavior of lower/upperUTF8 is undefined by definition. In this
-            /// case, the behavior is also reasonable.
-            size_t valid_size = output.size();
-            if (!output.empty() && output.back() == '\0')
-                --valid_size;
-
-            res_data.resize(curr_offset + valid_size + 1);
-
-            memcpy(&res_data[curr_offset], output.data(), valid_size);
-            res_data[curr_offset + valid_size] = 0;
-
-            curr_offset += valid_size + 1;
+                out_size = ucasemap_utf8ToLower(
+                    csm, reinterpret_cast<char *>(&res_data[curr_offset]), res_data.size() - curr_offset, data_start, size, &error_code);
+            // std::cout << size << ":" << out_size << ":" << static_cast<size_t>(res_data[curr_offset + out_size - 1]) << ":" << error_code
+            //   << std::endl;
+
+            if (error_code == U_BUFFER_OVERFLOW_ERROR)
+            {
+                size_t new_size = curr_offset + out_size + 1;
+                res_data.resize(new_size);
+
+                error_code = U_ZERO_ERROR;
+                if constexpr (upper)
+                    out_size = ucasemap_utf8ToUpper(
```

Comment:
Cosmetics: Let's remove the newlines here and in l. 84 and make the formatting similar as in l. 62/64.